### PR TITLE
[Messenger] Add more tests around the AMQP transport

### PR DESCRIPTION
**Pull request by sroze**
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ø
| License       | MIT
| Doc PR        | ø

Adding more tests to the AMQP transport/factory. These should have captured the following 3 bugs: #27198, #27197, #27196.

### CODE REVIEW

File: src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpTransportFactory.php [hunk 1]
```diff
@@ -34,7 +34,13 @@ public function __construct(EncoderInterface $encoder, DecoderInterface $decoder
 
     public function createTransport(string $dsn, array $options): TransportInterface
     {
-        return new AmqpTransport($this->encoder, $this->decoder, $dsn, $options, $this->debug);
+        return new AmqpTransport(
```
Comment (yceruto) [inline src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpTransportFactory.php]: According to the Code Standard shouldn't it be in one line?

Comment (fabpot) [reply src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpTransportFactory.php]: yes, it should

File: src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/AmqpTransportFactoryTest.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            true
+        );
+
+        $transport = $factory->createTransport('amqp://localhost', array('foo' => 'bar'));
+
+        $this->assertEquals(
+            new AmqpTransport($encoder, $decoder, Connection::fromDsn('amqp://localhost', array('foo' => 'bar'), true), array('foo' => 'bar'), true),
+            $transport
+        );
+    }
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/AmqpTransportFactoryTest.php]: should be on one line

File: src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/AmqpTransportTest.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        $decoder = $decoder ?: $this->getMockBuilder(DecoderInterface::class)->getMock();
+        $connection = $connection ?: $this->getMockBuilder(Connection::class)->disableOriginalConstructor()->getMock();
+
+        return new AmqpTransport(
+            $encoder,
+            $decoder,
+            $connection,
+            $options,
+            $debug
+        );
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/AmqpTransportTest.php]: one line as well

File: src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpTransport.php [hunk 4]
```diff
@@ -22,18 +22,17 @@ class AmqpTransport implements TransportInterface
 {
     private $encoder;
     private $decoder;
-    private $dsn;
+    private $connection;
     private $options;
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpTransport.php]: $options and $debug are not used anymore

**Comment by sroze** [comment]
@fabpot @yceruto @ogizanagi good catches, updated.

File: src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/AmqpTransportTest.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Messenger\Tests\Fixtures\DummyMessage;
+use Symfony\Component\Messenger\Transport\AmqpExt\AmqpTransport;
+use Symfony\Component\Messenger\Transport\AmqpExt\Connection;
+use Symfony\Component\Messenger\Transport\Serialization\DecoderInterface;
+use Symfony\Component\Messenger\Transport\Serialization\EncoderInterface;
+use Symfony\Component\Messenger\Transport\TransportInterface;
+
+class AmqpTransportTest extends TestCase
```
Comment (sroze) [inline src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/AmqpTransportTest.php]: Should requires amqp extension

**Review by ogizanagi — APPROVED** [review]

File: src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpTransport.php [hunk 6]
```diff
@@ -22,20 +22,15 @@ class AmqpTransport implements TransportInterface
...
 
-    public function __construct(EncoderInterface $encoder, DecoderInterface $decoder, string $dsn, array $options, bool $debug)
+    public function __construct(EncoderInterface $encoder, DecoderInterface $decoder, Connection $connection)
     {
         $this->encoder = $encoder;
         $this->decoder = $decoder;
-        $this->dsn = $dsn;
-        $this->options = $options;
-        $this->debug = $debug;
+        $this->connection = $connection;
```
[the remark this answers is not in the document]
Comment (sroze) [reply src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpTransport.php]: Non. The laziness is already [within the `Connection` object](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php#L162-L173). There is no point of making a class creation lazy 😄 (and even that is already lazy as behind service locators)

**Review by nicolas-grekas — APPROVED** [review]
Got it :)

**Pull request merged**